platform-sdk/docs/proposals/queuestate-fcqueue-to-virtualmap/proposal.md

```diff
@@ -0,0 +1,159 @@
+its next element has position N+1. When an empty queue is created, its head and tail are both
+pointing to 0. The head is _the next_ position to add, the tails is _the current_ position to
+remove. If head and tail positions are equal, the queue is empty.
+
+![Map keys](map-keys.svg)
+
+**Add**:
+
+* Insert an element to the head position
+* Increment the head by one
+* Code snippet:
+
+```java
+public void add(final E element) {
+    map.put(head++, element);
+}
+```
+
+![Add to the queue](queue-add.svg)
+
+**Poll / remove**:
+
+* Check if the queue is empty
+* Remove the element from the tail position
+* Increment the tail by one
+* Code snippet:
+
+```java
+public E remove() {
+    if (tail == head) {
+        return null;
+    }
+    return map.remove(tail++);
+}
+```
+
+![Remove from the queue](queue-remove.svg)
+
+### Serialization, deserialization, hashing, reconnects
+
+All provided by `VirtualMap`.
+
+`QueueNodeState` will write both its long values in addition to the label during serialization.
+
+### Data serialization and deserialization
+
+`FCQueue` today stores all data items in memory. When the queue is written to a snapshot, all
+items are serialized as `SelfSerializable`. To store items in `VirtualMap`, key and value
+serializers are to be provided. The key serializer will be trivial, as keys are longs. The value
+serializer will be implemented using the codec provided by `QueueNode`, similar to how key/value
+states work.
+
+## Data migration
+
+Current production states contain a few queue state nodes, all are tiny except transaction
+receipts. However, even the largest queue is expected to be small enough, so it can be migrated
+to a virtual map during Hedera node software update window.
+
+Migration code will be in `QueueNode.migrate()` This method will get its right child (which is
+`FCQueue`), iterate over it, insert all elements in the right order to a virtual map, then set
+the map as the new right child. The left child will be changed from a simple string leaf to a
+`QueueNodeState` leaf node.
+
+## Testing
+
+This feature doesn't add any new functionality. All existing Hedera unit tests should pass with
+the new queue state implementation. Unit tests specific to `QueueNode` may need to be updated.
+
+### Migration testing
+
+Migration testing will use a real mainnet state generated with a previous Hedera version. This
+state will be used to start a node with the latest version that includes queue state changes.
+
+## Performance impact
+
+**Queue operations (add, remove)**: both `FCQueue` and `VirtualMap` add and remove elements fast.
+No performance impact is expected.
+
+**Hashing**: `FCQueue` has a very efficient hashing algorithm, it hashes only newly inserted
+elements and "unhashes" removed elements. It's done on a single thread, however. `VirtualMap`
+needs to rehash more entities because of its tree structure, but it's done in parallel on
+many threads. No significant performance impact is expected.
```

Comment:
I'm unsure about the "no significant" part. You're saying the FCQueue's hashing algorithm is optimized to rehash the deltas. Additionally, the FCQueue lives in memory today. This sounds pretty fast to me.

On the other hand, you're saying the VM has to hash more entities (i.e. no support for optimized delta hashing), and on top of this, VMs live on disk. All of this sounds pretty slow to me.

So I'd expect some degradation of the hashing performance. It would be interesting to understand how large it would be.

Reply:
It's slightly more complicated than "in memory" vs "on disk". When a virtual map copy is hashed, all its dirty data (recently changed entities) is in the virtual node cache. Only after a copy is fully hashed, it can be flushed to disk. Some unchanged elements along the path to the root of the virtual tree may be loaded from disk, but not dirty leaf nodes. Given that virtual map is hashed on multiple threads, I assume there will be no significant impact on hashing time. Exact numbers can be provided only after this proposal is implemented.

Reply:
That still sounds a bit tricky, but it seems that this is because the hashing implementation in virtual maps isn't exactly trivial, which is fine.

One more question on this topic: why couldn't we add a third number to the `QueueNodeState` to facilitate delta hashing for the queue as a whole, basically retaining the existing hashing algorithm for queues? Could we consider this option in the design document and provide pros/cons to decide if we want to do that?

Reply:
It can be done, perhaps. I mean, maintain the running hash code at `QueueNode` level and ignore the hash code from the underlying virtual map. What's the benefit, though? This is extra hashing code, extra value to store somewhere, some additional tricks to disable virtual map hashing, etc.

Reply:
Sounds good.